### ISSUE TITLE
borg has mock as build dependency now

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#384]
likely due to the "get this working on readthedocs", we have a new (and somehow unexpected) build dependency now:

```
# I ran this in a fresh virtualenv:
(env)tw@tux:~/foo$ pip install borgbackup
Downloading/unpacking borgbackup
  Downloading borgbackup-0.28.0.tar.gz (396kB): 396kB downloaded
  Running setup.py (path:/home/tw/foo/env/build/borgbackup/setup.py) egg_info for package borgbackup
    
    Installed /home/tw/foo/env/build/borgbackup/setuptools_scm-1.8.0-py3.5.egg
    your setuptools is too old (<12)
    setuptools_scm functionality is degraded
    
    warning: no previously-included files matching '*.pyc' found under directory 'docs'
    warning: no previously-included files matching '*.pyo' found under directory 'docs'
    no previously-included directories found matching 'docs/_build'
    no previously-included directories found matching '.travis'
    warning: no previously-included files found matching '.coveragerc'
    warning: no previously-included files found matching '.gitattributes'
    warning: no previously-included files found matching '.gitignore'
    warning: no previously-included files found matching '.travis.yml'
    warning: no previously-included files found matching 'Vagrantfile'
Downloading/unpacking msgpack-python>=0.4.6 (from borgbackup)
  Downloading msgpack-python-0.4.6.tar.gz (116kB): 116kB downloaded
  Running setup.py (path:/home/tw/foo/env/build/msgpack-python/setup.py) egg_info for package msgpack-python
    
Installing collected packages: borgbackup, msgpack-python
  Running setup.py install for borgbackup
    your setuptools is too old (<12)
    setuptools_scm functionality is degraded
    calling custom build steps
    building 'borg.compress' extension
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include -I/usr/include -I/usr/local/include/python3.5m -c borg/compress.c -o build/temp.linux-x86_64-3.5/borg/compress.o
    gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/compress.o -L/usr/lib -L/usr/lib -llz4 -o build/lib.linux-x86_64-3.5/borg/compress.cpython-35m-x86_64-linux-gnu.so
    building 'borg.crypto' extension
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include -I/usr/include -I/usr/local/include/python3.5m -c borg/crypto.c -o build/temp.linux-x86_64-3.5/borg/crypto.o
    gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/crypto.o -L/usr/lib -L/usr/lib -lcrypto -o build/lib.linux-x86_64-3.5/borg/crypto.cpython-35m-x86_64-linux-gnu.so
    building 'borg.chunker' extension
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c borg/chunker.c -o build/temp.linux-x86_64-3.5/borg/chunker.o
    In file included from borg/chunker.c:241:0:
    borg/_chunker.c: In function ‘chunker_process’:
    borg/_chunker.c:237:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         while(c->remaining > c->window_size && ((sum & chunk_mask) || n < min_size)) {
                            ^
    borg/_chunker.c:237:69: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         while(c->remaining > c->window_size && ((sum & chunk_mask) || n < min_size)) {
                                                                         ^
    gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/chunker.o -o build/lib.linux-x86_64-3.5/borg/chunker.cpython-35m-x86_64-linux-gnu.so
    building 'borg.hashindex' extension
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c borg/hashindex.c -o build/temp.linux-x86_64-3.5/borg/hashindex.o
    In file included from borg/hashindex.c:239:0:
    borg/_hashindex.c: In function ‘hashindex_read’:
    borg/_hashindex.c:174:15: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         if(length != sizeof(HashHeader) + buckets_length) {
                   ^
    borg/_hashindex.c: In function ‘hashindex_write’:
    borg/_hashindex.c:278:54: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         if(fwrite(index->buckets, 1, buckets_length, fd) != buckets_length) {
                                                          ^
    gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/hashindex.o -o build/lib.linux-x86_64-3.5/borg/hashindex.cpython-35m-x86_64-linux-gnu.so
    building 'borg.platform_linux' extension
    gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c borg/platform_linux.c -o build/temp.linux-x86_64-3.5/borg/platform_linux.o
    gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/platform_linux.o -lacl -o build/lib.linux-x86_64-3.5/borg/platform_linux.cpython-35m-x86_64-linux-gnu.so
    auto-generating API documentation
    examining module borg/archiver.py
    examining module borg/upgrader.py
    examining module borg/__main__.py
    examining module borg/archive.py
    examining module borg/fuse.py
    examining module borg/platform.py
    examining module borg/locking.py
    examining module borg/repository.py
    examining module borg/lrucache.py
    examining module borg/remote.py
    examining module borg/__init__.py
    examining module borg/xattr.py
    examining module borg/helpers.py
    examining module borg/cache.py
    examining module borg/_version.py
    examining module borg/key.py
    examining module borg/logger.py
    examining module borg/platform_darwin.pyx
    examining module borg/platform_linux.pyx
    examining module borg/hashindex.pyx
    examining module borg/compress.pyx
    examining module borg/chunker.pyx
    examining module borg/crypto.pyx
    examining module borg/platform_freebsd.pyx
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/tw/foo/env/build/borgbackup/setup.py", line 297, in <module>
        install_requires=install_requires,
      File "/usr/local/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/local/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/tw/foo/env/lib/python3.5/site-packages/setuptools/command/install.py", line 59, in run
        return orig.install.run(self)
      File "/usr/local/lib/python3.5/distutils/command/install.py", line 539, in run
        self.run_command('build')
      File "/usr/local/lib/python3.5/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/usr/local/lib/python3.5/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/usr/local/lib/python3.5/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/tw/foo/env/build/borgbackup/setup.py", line 231, in run
        self.run_command('build_usage')
      File "/usr/local/lib/python3.5/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/home/tw/foo/env/build/borgbackup/setup.py", line 142, in run
        from borg.archiver import Archiver
      File "/home/tw/foo/env/build/borgbackup/borg/archiver.py", line 33, in <module>
        from .archive import Archive, ArchiveChecker, CHUNKER_PARAMS
      File "/home/tw/foo/env/build/borgbackup/borg/archive.py", line 29, in <module>
        import mock
    ImportError: No module named 'mock'
    generating usage docs
    Complete output from command /home/tw/foo/env/bin/python3.5 -c "import setuptools, tokenize;__file__='/home/tw/foo/env/build/borgbackup/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-rl8izl90-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/tw/foo/env/include/site/python3.5:
    your setuptools is too old (<12)

setuptools_scm functionality is degraded

running install

running build

running build_py

creating build

creating build/lib.linux-x86_64-3.5

creating build/lib.linux-x86_64-3.5/borg

copying borg/archiver.py -> build/lib.linux-x86_64-3.5/borg

copying borg/upgrader.py -> build/lib.linux-x86_64-3.5/borg

copying borg/__main__.py -> build/lib.linux-x86_64-3.5/borg

copying borg/archive.py -> build/lib.linux-x86_64-3.5/borg

copying borg/fuse.py -> build/lib.linux-x86_64-3.5/borg

copying borg/platform.py -> build/lib.linux-x86_64-3.5/borg

copying borg/locking.py -> build/lib.linux-x86_64-3.5/borg

copying borg/repository.py -> build/lib.linux-x86_64-3.5/borg

copying borg/lrucache.py -> build/lib.linux-x86_64-3.5/borg

copying borg/remote.py -> build/lib.linux-x86_64-3.5/borg

copying borg/__init__.py -> build/lib.linux-x86_64-3.5/borg

copying borg/xattr.py -> build/lib.linux-x86_64-3.5/borg

copying borg/helpers.py -> build/lib.linux-x86_64-3.5/borg

copying borg/cache.py -> build/lib.linux-x86_64-3.5/borg

copying borg/_version.py -> build/lib.linux-x86_64-3.5/borg

copying borg/key.py -> build/lib.linux-x86_64-3.5/borg

copying borg/logger.py -> build/lib.linux-x86_64-3.5/borg

creating build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/archiver.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/hashindex.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/upgrader.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/archive.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/platform.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/locking.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/chunker.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/repository.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/lrucache.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/crypto.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/benchmark.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/__init__.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/compress.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/xattr.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/helpers.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/key.py -> build/lib.linux-x86_64-3.5/borg/testsuite

copying borg/testsuite/logger.py -> build/lib.linux-x86_64-3.5/borg/testsuite

creating build/lib.linux-x86_64-3.5/borg/support

copying borg/support/argparse.py -> build/lib.linux-x86_64-3.5/borg/support

copying borg/support/__init__.py -> build/lib.linux-x86_64-3.5/borg/support

calling custom build steps

running build_ext

building 'borg.compress' extension

creating build/temp.linux-x86_64-3.5

creating build/temp.linux-x86_64-3.5/borg

gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include -I/usr/include -I/usr/local/include/python3.5m -c borg/compress.c -o build/temp.linux-x86_64-3.5/borg/compress.o

gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/compress.o -L/usr/lib -L/usr/lib -llz4 -o build/lib.linux-x86_64-3.5/borg/compress.cpython-35m-x86_64-linux-gnu.so

building 'borg.crypto' extension

gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include -I/usr/include -I/usr/local/include/python3.5m -c borg/crypto.c -o build/temp.linux-x86_64-3.5/borg/crypto.o

gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/crypto.o -L/usr/lib -L/usr/lib -lcrypto -o build/lib.linux-x86_64-3.5/borg/crypto.cpython-35m-x86_64-linux-gnu.so

building 'borg.chunker' extension

gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c borg/chunker.c -o build/temp.linux-x86_64-3.5/borg/chunker.o

In file included from borg/chunker.c:241:0:

borg/_chunker.c: In function ‘chunker_process’:

borg/_chunker.c:237:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

     while(c->remaining > c->window_size && ((sum & chunk_mask) || n < min_size)) {

                        ^

borg/_chunker.c:237:69: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

     while(c->remaining > c->window_size && ((sum & chunk_mask) || n < min_size)) {

                                                                     ^

gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/chunker.o -o build/lib.linux-x86_64-3.5/borg/chunker.cpython-35m-x86_64-linux-gnu.so

building 'borg.hashindex' extension

gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c borg/hashindex.c -o build/temp.linux-x86_64-3.5/borg/hashindex.o

In file included from borg/hashindex.c:239:0:

borg/_hashindex.c: In function ‘hashindex_read’:

borg/_hashindex.c:174:15: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

     if(length != sizeof(HashHeader) + buckets_length) {

               ^

borg/_hashindex.c: In function ‘hashindex_write’:

borg/_hashindex.c:278:54: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

     if(fwrite(index->buckets, 1, buckets_length, fd) != buckets_length) {

                                                      ^

gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/hashindex.o -o build/lib.linux-x86_64-3.5/borg/hashindex.cpython-35m-x86_64-linux-gnu.so

building 'borg.platform_linux' extension

gcc -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/include/python3.5m -c borg/platform_linux.c -o build/temp.linux-x86_64-3.5/borg/platform_linux.o

gcc -pthread -shared build/temp.linux-x86_64-3.5/borg/platform_linux.o -lacl -o build/lib.linux-x86_64-3.5/borg/platform_linux.cpython-35m-x86_64-linux-gnu.so

running build_api

auto-generating API documentation

examining module borg/archiver.py

examining module borg/upgrader.py

examining module borg/__main__.py

examining module borg/archive.py

examining module borg/fuse.py

examining module borg/platform.py

examining module borg/locking.py

examining module borg/repository.py

examining module borg/lrucache.py

examining module borg/remote.py

examining module borg/__init__.py

examining module borg/xattr.py

examining module borg/helpers.py

examining module borg/cache.py

examining module borg/_version.py

examining module borg/key.py

examining module borg/logger.py

examining module borg/platform_darwin.pyx

examining module borg/platform_linux.pyx

examining module borg/hashindex.pyx

examining module borg/compress.pyx

examining module borg/chunker.pyx

examining module borg/crypto.pyx

examining module borg/platform_freebsd.pyx

running build_usage

Traceback (most recent call last):

  File "<string>", line 1, in <module>

  File "/home/tw/foo/env/build/borgbackup/setup.py", line 297, in <module>

    install_requires=install_requires,

  File "/usr/local/lib/python3.5/distutils/core.py", line 148, in setup

    dist.run_commands()

  File "/usr/local/lib/python3.5/distutils/dist.py", line 955, in run_commands

    self.run_command(cmd)

  File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command

    cmd_obj.run()

  File "/home/tw/foo/env/lib/python3.5/site-packages/setuptools/command/install.py", line 59, in run

    return orig.install.run(self)

  File "/usr/local/lib/python3.5/distutils/command/install.py", line 539, in run

    self.run_command('build')

  File "/usr/local/lib/python3.5/distutils/cmd.py", line 313, in run_command

    self.distribution.run_command(command)

  File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command

    cmd_obj.run()

  File "/usr/local/lib/python3.5/distutils/command/build.py", line 135, in run

    self.run_command(cmd_name)

  File "/usr/local/lib/python3.5/distutils/cmd.py", line 313, in run_command

    self.distribution.run_command(command)

  File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command

    cmd_obj.run()

  File "/home/tw/foo/env/build/borgbackup/setup.py", line 231, in run

    self.run_command('build_usage')

  File "/usr/local/lib/python3.5/distutils/cmd.py", line 313, in run_command

    self.distribution.run_command(command)

  File "/usr/local/lib/python3.5/distutils/dist.py", line 974, in run_command

    cmd_obj.run()

  File "/home/tw/foo/env/build/borgbackup/setup.py", line 142, in run

    from borg.archiver import Archiver

  File "/home/tw/foo/env/build/borgbackup/borg/archiver.py", line 33, in <module>

    from .archive import Archive, ArchiveChecker, CHUNKER_PARAMS

  File "/home/tw/foo/env/build/borgbackup/borg/archive.py", line 29, in <module>

    import mock

ImportError: No module named 'mock'

generating usage docs

----------------------------------------
Cleaning up...
Command /home/tw/foo/env/bin/python3.5 -c "import setuptools, tokenize;__file__='/home/tw/foo/env/build/borgbackup/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-rl8izl90-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/tw/foo/env/include/site/python3.5 failed with error code 1 in /home/tw/foo/env/build/borgbackup
Storing debug log for failure in /home/tw/.pip/pip.log
```

**Comment by ThomasWaldmann** [comment]
@anarcat can you have a look?

**Comment by ThomasWaldmann** [comment]
Also: is it necessary to run build_api and build_usage? We can't build the docs anyway (or else we would have to require sphinx etc. also, which is a rather unwanted, big dependency).

**Comment by ThomasWaldmann** [comment]
fixed by:
https://github.com/borgbackup/borg/commit/3238b8ea76013a3d8a3b8d6089a31d34ab55ddb7

**Comment by anarcat** [comment]
boom, missed that, sorry. note that this should be triggered only when the C shims are not compiled/available..

**Comment by ThomasWaldmann** [comment]
I didn't understand your last comment. What are you referring to?

**Comment by anarcat** [comment]
I thought the mock stuff would kick in only if msgpack and our other pyx stuff was missing.

**Comment by LocutusOfBorg** [comment]
sorry but in Debian we provide a borgbackup and a borgbackup-doc package.

what about trying to detect the doc packages at runtime and then build the documentation accordingly?

that way I'm in a bad situation:
I can't build the documentation with msgpack, and if I try to build it
``` READTHEDOCS=1 ``` I can't rebuild c files:

```
if not on_rtd:
    ext_modules += [
    Extension('borg.compress', [compress_source], libraries=['lz4'], include_dirs=include_dirs, library_dirs=library_dirs),
    Extension('borg.crypto', [crypto_source], libraries=['crypto'], include_dirs=include_dirs, library_dirs=library_dirs),
    Extension('borg.chunker', [chunker_source]),
    Extension('borg.hashindex', [hashindex_source])
]
```
this is bad for us
(of course I can patch it to remove the if statement, but I'm unsure about the rationale behind doing that)

**Comment by ThomasWaldmann** [comment]
Seems like this needs another try...

**Comment by anarcat** [comment]
since mock is part of unittest and unittest part of the standard library, couldn't we just use that in this case?

**Comment by ThomasWaldmann** [comment]
mock is not part of stdlib as of python 3.2, so we can not do that now.

**Comment by anarcat** [comment]
so this whole nastiness was introduced in 80f82ef..1c61f87, for #262. i had to do this because usage fires up all borg commands, which would fail because the proper pyx libs are not installed just yet. this includes `msgpack`, but also `chunker`, `compress`, `crypto`, `hashindex` and `platform`. (`archiver.y` also disables `cache`, `key`,  `repository` and `upgrader`, but that's probably a mistake.)

as i mentionned in #262, [readthedocs](https://read-the-docs.readthedocs.org/en/latest/faq.html#i-get-import-errors-on-libraries-that-depend-on-c-modules) suggests using `mock` for exactly that purpose. i didn't exactly do it the right way (register fake modules in `sys.modules` instead of faking them one by one), but they also depend on `mock`. for `3.2`, they suggest having it as a dependency.

so I am not sure how to fix this: i could rewire a fake mock, but that way be a little too ironic. we could ship our own mock in `borg/support`, but that begs the question: why would it be a problem to have mock as a build depend? it's in the standard lib starting with 3.3 anyways...

**Comment by anarcat** [comment]
see #400 for a dumb fix.

**Comment by anarcat** [comment]
so i closed #400. after seeing where #208 is going, i think it makes more sense to commit the generated files into git. it will be a little more painful to maintain, and we are likely to forget some changes some times, but the maintenance burden here is getting way too ridiculous to justify the mess it has created.

so my current approach with this would be to:

1. drop all the mock and `have_cython()` crap from the tree completely
2. remove the API and usage builds from the main setup command
3. build and commit the `.rst` files for usage and API
4. add documentation on how to generate the rst files
5. for #208, add the examples and extra pieces to the usage files and do *not* include them directly (use a table of contents instead), followup there.

@ThomasWaldmann makes sense?

**Comment by anarcat** [comment]
so i did what i suggested in last comment in #408. step 5 can be handled by #208 directly.
